src/arch/riscv/pmp.cc

```diff
@@ -0,0 +1,252 @@
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "arch/riscv/pmp.hh"
+
+#include "arch/generic/tlb.hh"
+#include "arch/riscv/isa.hh"
+#include "base/addr_range.hh"
+#include "base/types.hh"
+#include "cpu/thread_context.hh"
+#include "math.h"
+#include "mem/request.hh"
+#include "params/PMP.hh"
+#include "sim/sim_object.hh"
+
+PMP::PMP(const Params &params) :
+    SimObject(params),
+    maxEntries(params.max_pmp)
+{
+    for (int i=0; i < maxEntries; i++) {
+        pmpEntry* entry = new pmpEntry;
+        AddrRange thisRange(-1,-2);
+        entry->pmpAddr = thisRange;
+        entry->pmpCfg = 0;
+        pmpTable.push_back(entry);
+    }
+
+    num_rules = 0;
+
+}
+
+bool
+PMP::pmpCheck(const RequestPtr &req, BaseTLB::Mode mode,
+              RiscvISA::PrivilegeMode pmode)
+{
+    // This function checks if the request
+    // is allowed based on PMP rules
+    if (num_rules == 0)
+        return true;
+
+    int ret = -1;
+
+    // all pmp entries need to be looked from the lowest to
+    // the highest number
+
+    int match_index = -1;
+
+    for (int i = 0; i < pmpTable.size(); i++) {
+        //std::cout << "pmp: table check " << i << "   " << std::endl;
+        if (pmpTable[i]->pmpAddr.contains(req->getPaddr())){
+           //address matched
+            match_index = i;
+        }
+
+        if ((PMP_OFF != pmpGetAField(pmpTable[match_index]->pmpCfg))
+                                            && (match_index > -1)){
+            // check the RWX permissions from the pmp entry
+            uint8_t allowed_privs = PMP_READ | PMP_WRITE | PMP_EXEC;
+
+            // i is the index of pmp table which matched
+            allowed_privs &= pmpTable[match_index]->pmpCfg;
+
+            if ((mode == BaseTLB::Mode::Read) && (PMP_READ & allowed_privs))
+                {ret = 1;
+                break;
+                }
+            else if ((mode == BaseTLB::Mode::Write) &&
+                                            (PMP_WRITE & allowed_privs))
+                {ret = 1;
+                break;
+                }
+            else if ((mode == BaseTLB::Mode::Execute) &&
+                                            (PMP_EXEC & allowed_privs))
+                {ret = 1;
+                break;
+                }
+            else{
+                ret = 0;
+                break;
+                }
+        }
+
+    }
+
+    // No PMP entry matched
+    if (ret == -1) {
+        if (pmode == RiscvISA::PrivilegeMode::PRV_M) {
```

Comment:
Move this check to the top. If we're in M, just return true.

If you do this, then you don't need `ret` or the `breaks`. Just return true/false if you have a match

Reply:
done